### DISCUSSION TITLE
Carelink login not working with xdrip

### DISCUSSION THREAD

**Discussion by iwascu** [NightscoutFoundation/xDrip#2821] [General]
the carelink server only works with medtronic applications. xdrip can no longer connect to the server. do you have the same problem? (server is carelink.minimed.eu)

**Comment by Geluboy** [comment]
I have this problem too. I mean login method changed. 
![inbound665463902367197569](https://user-images.githubusercontent.com/101523710/236905571-8165f381-079c-4353-9918-085507746282.png)

**Comment by Navid200** [comment]
@benceszasz

Do you have any info?
Thanks

**Comment by Geluboy** [comment]
About 1 mounth ago when i can't login I must create new account. Becouse was wrong type datas in Medtronic server. The guy who made Guardian monitor ios app he changed login method its worked, and working it.
Well then suggest:
"Then in the application they should add the configuration whether the sensor connects to the pump or through the application (without the pump) because the data-based detection mechanism does it wrong and the fact that "deviceFaimily" = "GUARDIAN" does not mean that it is a BLE connection or not . If you give it to them, they should take care of it. But I don't know if that's for sure. Unfortunately I haven't found any other way to detect this and what my application does is first try to fetch data for the pumps and then for the non-pump version.
what's important - if the downloaded data is zero, it's probably downloaded from the wrong address and then the algorithm tries to download the data using a different approach. Automatic detection is not possible. If you give it to the programmer, he should take care of it.

Previously, the api /monitor/data request returned a structure with one deviceFamily parameter, and on this basis, if the name contained ”BLE”, there was a different path for downloading data from carelink - a different address. Currently, two addresses with data are being checked. And if there is data for one of them, I remember this path and download it correctly every time.

Well who can programing xdrip i hope understand 😁

**Comment by iwascu** [comment]
I mention that I am on android without a pump. guardian 3. It worked until today at 14:48 when the medtronic server went down. After they started the server, the applications from carelink medronic work instead, xdrip does not.

**Reply by Navid200** [reply]
@iwascu 
Please click on "Create issue from discussion" in the right pane.

**Comment by Ausher2** [comment]
I'm having the same problem today. 770G working with MiniMed app, Carelink follower app working, xDrip returns Data Request Error Response Code: 400

**Comment by benceszasz** [comment]
HTTP 504 error (timeout) seems to be because CareLink refuses to answer the requests sent by xdrip. I think maybe Medtronic has put the browser headers used in the HTTP requests to banned list. Changing the relevant headers (User-Agent) seems to solve this problem and CareLink answers with HTTP 200.

**Reply by Navid200** [reply]
There is a claim that the US server CareLink is still working fine.
Is it possible that fixing the problem could possibly break things for the US server?

**Reply by Navid200** [reply]
@benceszasz Are you interested in opening a PR?
Do you need help for it?

**Comment by gginard** [comment]
Hello

i think i found the problem

In this file (line 519):
https://github.com/NightscoutFoundation/xDrip/blob/fbdead532d3d83fffcc8acc7d6b858981a087d11/app/src/main/java/com/eveningoutpost/dexdrip/cgm/carelinkfollow/client/CareLinkClient.java

Original
.addHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36");

Changed by:
.addHeader("User-Agent", "Mozilla/5.0 (iPad; CPU OS 7_0_6 like Mac OS X) AppleWebKit/537.51.1 (KHTML, like Gecko) Mobile/11B651");

I test in python system with this client:
https://github.com/ondrej1024/carelink-python-client

      self.__commonHeaders = {
            # Common browser headers
            "Accept-Language":"en;q=0.9, *;q=0.8",
            "Connection":"keep-alive",
            "sec-ch-ua":"\"Google Chrome\";v=\"87\", \" Not;A Brand\";v=\"99\", \"Chromium\";v=\"87\"",
            #"User-Agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36",
            "User-Agent":"Mozilla/5.0 (iPad; CPU OS 7_0_6 like Mac OS X) AppleWebKit/537.51.1 (KHTML, like Gecko) Mobile/11B651",
            "Accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"
          }

I changed the "User-Agent", and i solved y my python system.

Medtronic is filtrer by "User-Agent" ;-(

**Reply by tloczekt** [reply]
Yes, you are right, I changed the same part of code and python script is working well now. I think the value Mobile/11B65 is the most important thing - 
probably Medtronic  filter by browser type.....chrom/87 is tool old or should be a mobile agent (like Carelink Connect or Minimed App send its agent data)

**Reply by Ausher2** [reply]
For those of us non-programmer users, how do we fix this?

**Reply by CenGo86** [reply]
> For those of us non-programmer users, how do we fix this?

We'll have to wait unless one of the two can upload a compiled version here for us until the update for xdrip comes.

**Reply by heliax01** [reply]
I am having the same problem. Looking forward to a new compiled verion.

**Reply by CenGo86** [reply]
Now it works for me. Dont know why

**Reply by gginard** [reply]
Medtronic has put back the original User-Agent

**Comment by Navid200** [comment]
Is the problem solved for everyone?

Is anyone still experiencing the problem with official xDrip?

**Comment by Ausher2** [comment]
I am still having the same problem, yes.

Ann

**Comment by ntroup** [comment]
While running this updated version, I am still seeing:
**CareLink download error! Response code: 400**

**Reply by Navid200** [reply]
What updated version.  We have not created any new versions to address this yet.

**Reply by ntroup** [reply]
Sorry, good point.  I have been trying the version from https://github.com/benceszasz/xDripCareLinkFollower with a fix for the 504 error, but that didn't fix the 400 error which I was getting with both this and the other version.

I saw the PR (https://github.com/NightscoutFoundation/xDrip/pull/2822) that pull those changes into this version, but I don't believe it fixes everything entirely.

**Reply by Navid200** [reply]
Does the official xDrip system status page show anything that could identify you or your location?

**Reply by ntroup** [reply]
See below, but I have it configured for 'United States'.  I do not see a server location.

**Comment by Navid200** [comment]
I would like to create a summary so that we may create a plan.  I need to know if you are still having problems or not and I need to know what the CareLink country is in your xDrip settings.  If you don't feel like sharing that, it is your choice and you don't have to.
My hope is to find what's common between those who don't have the problem any longer, and what is common between those who do.
We cannot solve a problem before we understand the cause perfectly.

If you tell me if you still have the problem or not and also what your CarLink country is, I will put everything in a single table to make it easier for everyone to see the summary.

**Reply by Navid200** [reply]
I myself am not a CareLink user.  Is there a setting on the status page that just shows a server location instead of the country?  If yes, you can just tell me the server location.

**Reply by ntroup** [reply]
Currently running nightly build:  **e50dfdf-2023-0503**
Carelink Country:  **United States (en)**

**Reply by Ausher2** [reply]
I am running the same build from the US.

**Reply by CenGo86** [reply]
Newest Nightly. Germany and no problems anymore since 9:00AM

**Reply by solanix** [reply]
I have the exact same issue. Started Midnight CST MAY 8, 2023.
US(en)       2023.05.03.

**Reply by Navid200** [reply]
@solanix One post is enough stating the same thing.  Also, you can edit a post if you realize after posting that you forgot to add something.

**Reply by kevinfields777** [reply]
Problem persists, United States (en) version 344deb4-2022.11.16

**Reply by solanix** [reply]
Has there been any progress or is there a plan as of yet? What about updating the user string as noted above? There are sites that block older browsers for security purposes and that shouldn't negatively impact any locale. What do you think? Can we try a RC that us in the US can try and report back?

**Reply by Navid200** [reply]
This feature was added to xDrip by @benceszasz.
I reached out to him in this same thread.  But, he has not replied to my questions.
It would be best if he tested and opened a PR.

We can blindly open a PR and merge (none of the active developers use CareLink).  But, that's not how we have operated so far.  
What do we do the next time it stops working?

**Reply by solanix** [reply]
Understood. I have been using xDrip to put data from Carelink primarily to display on my Garmin smartwatch. The predictive calculations have been fantastic. 
I am a retired software engineer and am willing to test on my end. I started this morning to setup the dev environment but have yet to have time to complete. I would suggest that the user agent string be a variable such that code is not necessarily required to be changed. It is unfortunate, but I have seen user agents been blocked numerous times due to aging out and security vulnerabilities with the browsers. Additionally, numerous tools use older user agent strings for nefarious reasons.

Additionally, a single PR is probably not a good idea here as there are two issues. The 504 and the 400.

I will test with the Python script and see what I get from the US.

**Reply by benceszasz** [reply]
@Navid200 sorry for not replying, but I am very busy with solving the "US issue". The situation is as follows. Medtronic released a new CareLink version after which the CareLink servers didn't respond to the HTTP requests used in xdrip. By changing the HTTP headers of the browser emulated logon procedure to newer browser headers CareLink servers started to respond again and everything was working again outisde the US. After some time (half / one day) CareLink EU server started to accept the older HTTP client headers again, but not the US server. It turned out, that Medtronic activated a new feature: multiple follow option. This new feature allows a carepartner to follow more than 1 patient (currently the limit is 5 patients if I remember correctly). But this new feature uses new CareLink APIs: retrieving the list of the patients, retrieving patient device family and also to get the current data of a patient. Even if a carepartner follows only 1 patient, if the CareLink Server is configured to support multiple patient following, then the new APIs must be used. I am working on to support this multiple patient follow: retrieve CareLink server configuration (is multiple follow enabled) and use the new APIs if multiple follow function is enabled on the CareLink server. So new logic, new APIs, new optional parameter (patient username), the old logic must remain as an alternative method for single patient follow CareLink servers.

**Reply by solanix** [reply]
@benceszasz  Your memory serves you correct. 5 patients. 

Is there something that I can assist with?

**Reply by Navid200** [reply]
@benceszasz Thanks for working on this.  Please tell us if you need help.

**Reply by benceszasz** [reply]
@solanix You can test the first version of my carelink xdrip client for multi patient follow. It should work with patient and carepartner US accounts too now. I could only test it with standalone Medtronic CGM, but not with 7XXG pump.

**Reply by solanix** [reply]
@benceszasz Installed and tested xdrip-carelink.zpk as well as xdrip-variantX.apk. The following screenshot was taken to detail the logs. 

- The results are the same with a set patient or none provided

![Screenshot_20230513_183535_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/2350242/f8a9f56a-3891-4b5a-9348-9ff8a0f6e54b)

**Reply by benceszasz** [reply]
I assume you are using 770G. As I wrote earlier I could only test it first with standalone CGM and both Patient and Care Partner accounts were working fine. It turned out, that 7xxG pumps in the US won't work with the same data APIs even on the new multiple follow enabled servers. Currently I have no solution for US 770G pump Care Partner accounts only for Patient accounts.
I have released a new version, which should work in the US with 770G pumps but only using **Patient** accounts.

**Reply by solanix** [reply]
Thank you for all of the work and the rapid work around. I will give that a try.

Yes with regard to the 770g, G3 CGM, and Samsung Note 20 Ultra. I was using a follow account as well.

**Comment by Navid200** [comment]
| User | Problem persists | CareLink Country | xDrip version |  
| ----- | ------------------ | ------------------- | --------------- |  
| ntroup | Yes | United States (en) | 2023.05.03 |  
| Ausher2 | Yes | United States (en) | 2023.05.03 |  
| solanix | Yes | United States (en) | |  
| kevinfields777 | Yes | United States (en) | 2022.11.16 |  
| CenGo86 | No | Germany | 2023.05.03 |

**Comment by CenGo86** [comment]
I want to share something new. Im in germany and xdrip works fine. But I got message data processing error. This is the log. But everything works.

I don't know if this is related to this topic, but maybe it will help

![Screenshot_20230514_111029_xDrip+.jpg](https://github.com/NightscoutFoundation/xDrip/assets/86983057/f75a4331-7bd5-4b09-8306-cc0ddb9baa59)

**Comment by Ausher2** [comment]
I tried all of the variants and still got the same "data request  failed" error message each time. I logged out of Carelink and back in again but that didn't help. 

![Screenshot_20230514-073822](https://github.com/NightscoutFoundation/xDrip/assets/36939359/7103bab9-be4e-4535-b2e0-61019a840bee)

**Reply by benceszasz** [reply]
My latest private xdrip version should work with any account type and any device type.

**Comment by Geluboy** [comment]
Bence Szász version working for me in Hungary! Thx

**Comment by solanix** [comment]
This is aweful. I was wondering why I kept getting login failed and then a slew of two-factor authentication emails.

Is there a way to disable the MFA on the carelink account? I have not seen a toggle anywhere and it shows active under My Profile. I am hoping that this is just for some accounts and not an automatic enrollment across the US

**Reply by benceszasz** [reply]
Not sure, if it can be disabled. Meanwhile I have released a new version, which should work with any account type, so if there is no other way, you can create a Care Partner account without MFA and use it to follow your Patient account.

**Reply by solanix** [reply]
My hat's off to you brother. That is a fantastic drop that works as advertised. Thank you for your work on this

**Comment by Ausher2** [comment]
Thank you so much! It's so nice to have it working again on my Pixel Watch. I appreciate all your efforts.

**Comment by jamorham** [comment]
@benceszasz will you prepare a PR for this? thanks

**Reply by benceszasz** [reply]
@jamorham yes, I will, but I need some time. Meanwhile as far as I can tell the current xdrip CareLink Follower should work with Patient account in every region, only Care Partner accounts are not working in the US.

**Reply by Navid200** [reply]
@blacktails11 
This repository and all of its resources, like this forum, are for keeping the community together.
If you want to use someone else's  project, we cannot stop you.
However, no links to other repositories should be posted in these forums.

Everyone is welcome to contribute to this repository to make it stronger.  
Everyone can also choose to create their own version trying to break up the community.  We cannot stop people from doing that.  But, it is my belief that we are stronger when we stand together.
Would you please not use this forum to do it?

Would you please delete the links you have posted to the other repository?

As @benceszasz already explained in this same thread, all you need to do, using official xDrip, is to use the patient account instead of the care partner account.

**Reply by ghost** [reply]
@Navid200 thanks for your message.

Like i mentioned in my first message, I don't even know what Github is....
Nevermind a repositorie but i do know what links are so i will delete them

I just wanted to say thank you is all :)

**Reply by Navid200** [reply]
@blacktails11 Please delete both of your posts.  Or, I can delete them for you.

You can use xDrip using the patient account to use CareLink.

**Reply by ghost** [reply]
> @blacktails11 
> You can use xDrip using the patient account to use CareLink.

Thanks Navid.... I dont know what that means.....

**Reply by ghost** [reply]
Thanks for the welcome @Navid200 
Guess im not welcome here

**Reply by Navid200** [reply]
@blacktails11 What do you mean you are not welcome here?
Everyone is welcome here.  I asked you to please delete references to other repositories.  You deleted one link and left the references in.  I asked you to delete the posts.  You ignored my request and asked another question.  
Every time we post here, everyone in this thread gets a notification.

You also stated that it was working for you.  If your problem has been resolved, and you have any other questions, would you please open another thread?

**Comment by benceszasz** [comment]
@ntroup @Ausher2 @solanix @kevinfields777 have you all been using CareLink Follower with a Care Partner account in the US? If so, have you tried using the Patient account?

**Reply by ntroup** [reply]
I've been using the Care Partner account, as my Patient account has Multi-Factor enabled and I can't turn it off.

**Reply by steinauf85** [reply]
> I've been using the Care Partner account, as my Patient account has Multi-Factor enabled and I can't turn it off.

Same with me, though I’ve seen on Facebook others that resolved this by switching to the patient account through the MiniMed Mobile app. 

Edit: never mind, just saw the latest update which fixed it. Thank you @benceszasz !

**Reply by kevinfields777** [reply]
Sorry for the late reply, for some reason I couldn't access the link to the discussion until now.  I switched from the partner account to patient account and it started working again.

**Comment by Navid200** [comment]
Please try the latest Nightly:  https://github.com/NightscoutFoundation/xDrip/releases/tag/2023.05.26

**Reply by djgingersteve** [reply]
Thanks. Fixed for me in United Kingdom on a Partner account 🙏🏻

**Comment by Navid200** [comment]
@benceszasz Thanks for the fix

**Comment by aruiz** [comment]
Is this issue fixed in the official releases?

**Comment by Carmoldu** [comment]
Hello,
 Today the carelink server went down for a while (at least in Spain). When it came back up, the login functionality did not work: it would take you to the login webpage and after login it would take you back to the Home screen of xdrip+ where nothing would happen. After a few seconds, the message "Not logged in! Please log in!" would appear. Sometimes after the login, a message stating xdrip+ stopped working would appear.

This sounds a lot like the issue mentioned in this conversation. Should the issue be re-opened?

**Reply by Carmoldu** [reply]
Benceszasz's variant seems to work though.

**Reply by Geluboy** [reply]
Well we get a messenges, was a big medtronic server update. Not worked it hungaryan time 17:00-23:00... 
Now working for me, but I use Bence Szász version not the official xdrip.

**Reply by sarhatabaot** [reply]
This is happening for us again as well.

**Reply by kevinfields777** [reply]
I had this issue a week ago, upgrade to the latest official nightly release.

**Comment by tloczekt** [comment]
This version is working well
https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.03.27?fbclid=IwAR1UP4PRCldC4ylXLrOPdZJcjVt9ISwEZcWgFskzYar1wyhdn0nHgPsIPAs

niedz., 7 kwi 2024, 07:01 użytkownik Omer Oreg ***@***.***>
napisał:

**Reply by kevinfields777** [reply]
Confirmed.

**Comment by benceszasz** [comment]
The latest CareLink Follower version (released on 05.12.2023) uses a much better login method than the previous versions:
https://github.com/NightscoutFoundation/xDrip/pull/3210
Update your xDrip at least to 05.12.2023 or a later Nightly version:
https://github.com/NightscoutFoundation/xDrip/releases
The newer version works fine: I have checked after the CareLink update with different account types, devices and regions, all works!
